src/main/scala/fi/oph/koski/validation/KoskiValidator.scala

```diff
@@ -358,6 +359,46 @@ class KoskiValidator(tutkintoRepository: TutkintoRepository, val koodistoPalvelu
     }
   }
 
+  private def validateOsaAikainenErityisopetus(opiskeluoikeus: KoskeenTallennettavaOpiskeluoikeus) = {
+    (opiskeluoikeus, opiskeluoikeus.lisätiedot) match {
+      case ((o: EsiopetuksenOpiskeluoikeus), Some(l: EsiopetuksenOpiskeluoikeudenLisätiedot)) => validateEsiopetuksenOsaAikainenErityisopetus(o, l)
+      case ((o: PerusopetuksenOpiskeluoikeus), Some(l: PerusopetuksenOpiskeluoikeudenLisätiedot)) => validatePerusopetuksenOsaAikainenErityisopetus(o, l)
+      case ((o: PerusopetuksenLisäopetuksenOpiskeluoikeus), Some(l: PerusopetuksenOpiskeluoikeudenLisätiedot)) => validatePerusopetuksenLisäopetuksenOsaAikainenErityisopetus(o, l)
+      case _ => HttpStatus.ok
+    }
+  }
+
+  private def validateEsiopetuksenOsaAikainenErityisopetus(o: EsiopetuksenOpiskeluoikeus, lisätiedot: EsiopetuksenOpiskeluoikeudenLisätiedot) = {
+    // TODO
+    HttpStatus.ok
+  }
+
+  private def validatePerusopetuksenOsaAikainenErityisopetus(o: PerusopetuksenOpiskeluoikeus, lisätiedot: PerusopetuksenOpiskeluoikeudenLisätiedot) = {
+    // TODO
+    HttpStatus.ok
+  }
+
+  private def validatePerusopetuksenLisäopetuksenOsaAikainenErityisopetus(o: PerusopetuksenLisäopetuksenOpiskeluoikeus, lisätiedot: PerusopetuksenOpiskeluoikeudenLisätiedot) = {
+    val tukimuodoissaOsaAikainenErityisopetus = sisältääOsaAikaisenErityisopetuksen(lisätiedot.erityisenTuenPäätökset) || sisältääOsaAikaisenErityisopetuksen(lisätiedot.tehostetunTuenPäätökset)
+    val suorituksessaOsaAikainenErityisopetus = o.suoritukset(0).osaAikainenErityisopetus
+
+    (tukimuodoissaOsaAikainenErityisopetus, suorituksessaOsaAikainenErityisopetus) match {
+      case (true, false) => KoskiErrorCategory.badRequest.validation.osaAikainenErityisopetus.perusopetuksenLisäopetuksenKirjausPuuttuu()
+      case _ => HttpStatus.ok
+    }
+  }
+
+  private def sisältääOsaAikaisenErityisopetuksen(t: Option[List[Tukimuodollinen]]) = {
+    // TODO: Tähän on varmaan joku scalampi tapa?
```

Comment:
Jos määrittelisi `Tukimuodollinen`-traittiin metodin
```
def tukimuotoLista: List[Koodistokoodiviite] = tukimuodot.toList.flatten
```
Niin ton voisi kirjoittaa näin
```
  private def sisältääOsaAikaisenErityisopetuksen(t: Option[List[Tukimuodollinen]]): Boolean = {
    val tukimuodot = t.flatMap(_.flatMap(_.tukimuotoLista)).map(_.koodiarvo).toList
    tukimuodot.contains("1")
  }
```

Reply:
Tuo koodi ei sellaisenaan käänny, vaikka esim. IDEA:n editori ei virhettä huomaakaan. Ongelma on varmaan se, että flatMap ei toimi ihan noin intuitiivisesti Option(List(...)):llä.

Päädyin nyt koodiin
```
  def tukimuotoLista: List[Koodistokoodiviite] = tukimuodot.getOrElse(List())

  
val tukimuodot = t.getOrElse(List()).flatMap(_.tukimuotoLista).map(_.koodiarvo).toList
```
, joka on myös mielestäni selkeämpi, kun ei ole sisäkkäisiä kutsuja. Tykkään myös tuosta getOrElse:stä, kun se kertoo vähän eksplisiittisemmin, mitä halutaan.